File: adapters/react-native/android/src/main/java/com/sandrox/levixel/reactnative/LevixelModule.kt
package com.sandrox.levixel.reactnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Coroutine

class LevixelModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("Levixel")

        View(LevixelView::class) {
            Events("onSourcePress", "onViewerEvent", "onControllerReady")
            // Coroutine view functions run on the UI thread in both RN architectures.
            // Promise-style view functions can wait for an unrelated UI batch in Paper.
            AsyncFunction("open") Coroutine { view: LevixelView, options: Map<String, Any?> -> view.open(options) }
            AsyncFunction("close") Coroutine { view: LevixelView -> view.close() }
            AsyncFunction("retry") Coroutine { view: LevixelView -> view.retry() }

            Prop("isController") { view: LevixelView, isController: Boolean ->
                view.isController = isController
            }
            Prop("items") { view: LevixelView, items: Array<Map<String, Any?>> ->
                view.items = items.toList()
            }
            Prop("index") { view: LevixelView, index: Int ->
                view.initialIndex = index
            }
            Prop("galleryId") { view: LevixelView, galleryId: String ->
                view.galleryId = galleryId
            }
            Prop("sourceId") { view: LevixelView, sourceId: String ->
                view.sourceId = sourceId
            }
            Prop("sourceCornerRadius") { view: LevixelView, sourceCornerRadius: Double ->
                view.sourceCornerRadius = sourceCornerRadius.toFloat()
            }
            OnViewDidUpdateProps { view: LevixelView ->
                view.commitSourceBinding()
            }
        }
    }
}

File: adapters/react-native/ios/LevixelModule.swift
import ExpoModulesCore

public final class LevixelModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Levixel")

        View(LevixelView.self) {
            Events("onSourcePress", "onViewerEvent")
            AsyncFunction("open") { (view: LevixelView, options: [String: Any], promise: Promise) in
                view.open(options: options, promise: promise)
            }
            AsyncFunction("close") { (view: LevixelView, promise: Promise) in view.close(promise: promise) }
            AsyncFunction("retry") { (view: LevixelView) in view.retry() }

            Prop("items") { (view: LevixelView, items: [[String: Any]]?) in
                view.items = items ?? []
            }
            Prop("index") { (view: LevixelView, index: Int?) in
                view.initialIndex = index ?? 0
            }
            Prop("galleryId") { (view: LevixelView, galleryId: String?) in
                view.galleryId = galleryId ?? ""
            }
            Prop("sourceCornerRadius") { (view: LevixelView, sourceCornerRadius: Double?) in
                view.sourceCornerRadius = CGFloat(sourceCornerRadius ?? 0)
            }
            OnViewDidUpdateProps { view in
                view.configureSourceView()
            }
        }
    }
}
